### PR TITLE
EPIC-MLOps-02: Residential Data Vaulting & MVP Ingestion

### LINKED ISSUE #53: [EPIC-MLOps-02] Data Engineering & Residential MVP Vaulting
# Issue: [EPIC-MLOps-02] Data Engineering & Residential MVP Vaulting

## Context & Objective
To build a production-grade **Residential Security System** using VideoMAE v2, we require a robust, reproducible data pipeline. The current objective is to ingest a surgical subset of the NTU RGB+D dataset that mirrors a residential courtyard environment, focusing specifically on interpersonal violence while minimizing false positives from normal interactions.

We must solve the "Large File Problem" by ensuring that heavy video binaries are never committed to Git, while maintaining strict version control for reproducibility.

### 1. Provision Immutable Data Vault
* Scaffold a structured directory hierarchy (`data/raw`, `data/processed`, `data/external`).
* Ensure the repository is configured to handle large binary payloads without impacting SCM performance.

### 2. Develop Surgical Extraction Pipeline
* Create an automated script (`extract_vault.py`) to process the NTU RGB+D `.zip` archives.
* **Flattening:** The script must extract files into a flat directory structure to optimize PyTorch I/O and simplify metadata parsing.
* **Sanitization:** Implement logic to strip hidden OS metadata (e.g., `__MACOSX`) and handle potentially corrupted archives.

### 3. Implement Residential Domain Logic
* Define a taxonomy for **Threat Vectors** (Punching, Pushing, Kicking).
* Define a taxonomy for **Hard Negatives** (Hugging, Waving, Walking).
* **Specific Constraint:** Class `A024` (Kicking something) must be treated as a Hard Negative to prevent false positives from sports or environmental interaction in a residential courtyard.

### 4. Provision MLOps Versioning (DVC)
* Initialize **Data Version Control (DVC)** within the hermetic project environment.
* Configure Git to ignore raw video directories while tracking the corresponding `.dvc` sidecar files.
* Establish a "Single Source of Truth" for the dataset that allows team members to replicate the environment exactly.

---

## Definition of Done (DoD)
- [ ] No `.avi` or `.zip` files are tracked by Git.
- [ ] Extraction script successfully filters target classes from the full NTU set.
- [ ] `data/raw.dvc` file is generated and contains the cryptographic hash of the current MVP subset.

## Required Stack
* **Dependencies:** `torch`, `torchvision`, `decord`, `dvc`, `pathlib`
* **Environment:** Hermetic Poetry Virtual Env
* **Hardware Target:** NVIDIA RTX 4090 (24GB VRAM)

### PR DESCRIPTION
**Pull request by mauriziokeningen**
# EPIC-MLOps-02: Residential Data Vaulting & MVP Ingestion

## Executive Summary
This PR establishes the **Data Engineering Foundation** for our Residential Security System. We have successfully implemented a hermetic data pipeline that surgically extracts a 13-class MVP from the NTU RGB+D dataset. 

By integrating **Data Version Control (DVC)**, we have solved the "Large File Problem" in Git, allowing us to manage gigabytes of video data via lightweight cryptographic hashes.

### 1. Residential Domain Adaptation
In a sterile hospital environment, property interaction (e.g., kicking a chair) is an anomaly. In a **Residential Courtyard**, it's a false positive (e.g., a child kicking a soccer ball). 
* **Decision:** Reclassified `A024` (Kicking something) as a **Hard Negative**. 
* **Impact:** The model is now explicitly trained to ignore high-velocity leg movements unless they intersect with another human bounding box (`A051`).

### 2. The "Flat Lake" Paradigm
Traditional hierarchical folder structures (`/class/subject/video.avi`) create computational overhead for PyTorch during recursive walks.
* **Decision:** Flattened all extractions into `data/raw/`.
* **Impact:** The Dataloader now performs a single $O(1)$ lookup using filename-string parsing to extract metadata (Subject, Action, Camera), maximizing I/O performance on our NVMe storage.

### 3. DVC Hybrid-SCM Architecture

We have decoupled the **Logic** (Git) from the **Payload** (DVC). 
* **Decision:** Initialized DVC in `--subdir` mode within the `research/` directory.
* **Impact:** Reviewers can pull code changes instantly without waiting for 5GB of binary data to download unless they explicitly trigger the data sync.

## 🛠️ Implementation Details
* **Surgical Extraction:** `scripts/extract_vault.py` filters 2,880 source files down to 624 targeted MVP samples.
* **High-Performance Decoding:** Integrated `decord` as our primary video backend. It utilizes C++ bindings to provide $3\times$ faster frame seeking compared to standard OpenCV.
* **Hermetic Locking:** Updated `pyproject.toml` with the full vision stack (`torch`, `torchvision`, `decord`, `dvc`).

---

## Testing & Onboarding Guide (Step-by-Step)

If this is your first time working with DVC in this repo, follow these instructions to enter the "Pit of Success."

### 1. Synchronize the Hermetic Environment
Ensure you have the required CLI tools (including DVC) without touching your global Python path:
```bash
git checkout epic/mlops02-data-vault
poetry install
```

### 2. Manual Payload Acquisition (MVP Phase)
*Since Cloud Remote Storage (S3) is scheduled for Epic 4, we must manually seed the vault:*
1. Place `nturgbd_rgb_s001.zip` into `research/data/raw/`.
2. Run the extraction pipeline:
   ```bash
   poetry run python scripts/extract_vault.py
   ```

### 3. Verify Data/Logic Decoupling
Confirm that the 624 extracted videos are being tracked by DVC and ignored by Git:
```bash
git check-ignore -v data/raw/S001C001P001R001A051_rgb.avi
# Expected: Output shows the file is ignored by .gitignore
```
---

## Quality Assurance Checklist
- [x] No binary data (`.avi` or `.zip`) has been committed to Git.
- [x] Extraction script handles "BadZip" errors and cleans hidden OS metadata.
- [x] DVC hashes are staged in `data/raw.dvc`.
- [x] Absolute pathing implemented via `pathlib` for cross-OS compatibility.

***

### Next Steps
Upon approval and merge of this PR, we will commence **[EPIC-MLOps-03]: VideoMAE v2 Training & Hyperparameter Tuning.**

### CODE REVIEW

**Comment by mauriziokeningen** [comment]
# PR: feat(data): Implement Resilient Sequential Acquisition Layer

## Executive Summary
This PR introduces `download_daemon.sh`, a fault-tolerant ingestion engine designed to decouple the **Data Acquisition** phase from active developer sessions. Given the high-latency egress from NTU servers and the ~250GB scale of the RGB+D dataset, this script provides the infrastructure necessary for background, multi-day synchronization without risking data corruption or terminal-hang failures.

## Architectural Improvements
* **Idempotency & Resilience:** Leverages `curl -C -` to enable byte-level resumption of interrupted transfers. This mitigates the risk of partial file corruption during long-haul network blips.
* **Operational Safety:** Implements `set -euo pipefail` (Strict Mode) to ensure the daemon terminates immediately upon critical failures rather than entering an undefined state.
* **Persistent Execution:** Architected to run within a `tmux` or `nohup` buffer, detaching the I/O process from the SSH lifecycle.

---

## Operational Guide

To initiate the ingestion pipeline, follow the standard persistent session protocol:

1.  **Initialize the Session:**
    ```bash
    tmux new -s data-vault
    ```
2.  **Export Credentials & Metadata:**
    (See the **Security & Mapping** section below for details).
    ```bash
    export NTU_SESSION_ID="your_token_here"
    export NTU_USER_AGENT="your_user_agent_here"
    ```
3.  **Execute the Daemon:**
    ```bash
    ./download_daemon.sh
    ```
4.  **Detach:**
    Press `Ctrl + B`, then let go and press `D`.

---

## Security & Mapping Management

This script follows a **Zero-Secret Policy**. Credentials and mapping logic are decoupled from the source code to ensure security and flexibility.

### 1. How to Obtain Ephemeral Credentials
* **Session ID:** 1. Log into the [NTU RGB+D Portal] (https://rose1.ntu.edu.sg/dataset/actionRecognition/download)](https://rose1.ntu.edu.sg/dataset/actionRecognition/download).
    2. Open **Developer Tools (F12)** -> **Application/Storage** -> **Cookies**.
    3. Copy the value of the `sessionid` key.
* **User-Agent String:** 1. In the same Developer Tools window, click the **Console** tab.
    2. Execute `navigator.userAgent`.
    3. Copy the resulting string. *Note: Matching the User-Agent to your session prevents anti-bot flagging.*

### 2. How to Identify Dataset Payload IDs
The script requires specific remote IDs to target data chunks. 
1. Navigate to the NTU Download page.
2. Hover your mouse over the **Download** button for a specific chunk (e.g., "nturgbd_rgb_s001.zip").
3. Right click and Inspect the URL.
4. The integer at the end (e.g., `.../download/126`) is your **Payload ID**.
5. **Current Mapping:** IDs 126 through 141 represent the sequential RGB chunks (S002-S017) required for our baseline.

> [!CAUTION]
> **Token Expiration:** NTU sessions typically rotate every 24–48 hours. If logs return a `403 Forbidden`, refresh the `NTU_SESSION_ID` environment variable and restart the daemon. The script will automatically resume from the last successful byte.

---

## Deployment Checklist
- [x] Verified `curl` byte-range support on target server (Accept-Ranges: bytes).
- [x] Validated strict mode failure handling (`set -euo pipefail`).
- [x] Confirmed directory auto-provisioning (`mkdir -p`).
- [x] Zero hardcoded secrets in `download_daemon.sh`.
- [x] Validated mapping of IDs 126–141 to local S002–S017 filenames.

**Review by JDavidOG — APPROVED** [review]
